src/core/inductive.ml

```diff
@@ -0,0 +1,257 @@
+          if s == ind_sym then
+            prf_of_p (List.map lift ts) (app (_Symb cons_sym) (List.rev xs))
+          else
+            fatal pos "%a is not a constructor of %a"
+              pp_symbol cons_sym pp_symbol ind_sym
+      | (Prod(a,b), _) ->
+          let (x,b) = Basics.unbind_name b in
+          let b = aux ((Bindlib.box_var x)::xs) b in
+          begin
+            match Basics.get_args a with
+            | (Symb(s), ts) ->
+                let b =
+                  (* In case of a recursive argument, we create an inductive
+                     hypothesis. *)
+                  if s == ind_sym then
+                    _Impl (prf_of_p (List.map lift ts) (Bindlib.box_var x)) b
+                  else b
+                in
+                _Prod (lift a) (Bindlib.bind_var x b)
+            | _ ->
+                fatal pos "The type of %a is not supported" pp_symbol cons_sym
+          end
+      | _ -> fatal pos "The type of %a is not supported" pp_symbol cons_sym
+    in
+    let res = aux [] !(cons_sym.sym_type) in
+    if !log_enabled then
+      log_ind "The clause of the constructor [%a] is %a"
+        Print.pp_term (Symb(cons_sym))
+        Print.pp_term (Bindlib.unbox res);
+    res
+  in
+
+  (* STEP 3: Create the conclusion of the inductive principle *)
+  let codom ts =
+    let x = Bindlib.new_var mkfree "x" in
+    let t = Bindlib.bind_var x (prf_of_p ts (_Vari x)) in
+    _Prod (fapp ind_sym ts) t
+  in
+  let conclusion = change_codom_ind_typ pos ind_sym codom in
+
+  (* STEP 4: Create the induction principle *)
+
+  let t =
+    List.fold_right (fun a b -> _Impl (case_of a) b) cons_list conclusion
+  in
+  let t = _Prod p_type (Bindlib.bind_var p t) in
+  if !log_enabled then
+    log_ind "The induction principle of the inductive type [%a] is %a"
+      Print.pp_term (Symb(ind_sym)) Print.pp_term (Bindlib.unbox t);
+  Bindlib.unbox t
+
+(** [type_to_pattern t] creates a pattern which matches with the type [t]. *)
+let rec type_to_pattern : term -> p_term = fun t ->
+  match t with
+  | Vari x    -> Pos.none (P_Patt (Some (Pos.none (Bindlib.name_of x)), None))
+  | Symb s       -> Pos.none (P_Iden (Pos.none ([], s.sym_name), true))
+  | Appl(t1, t2) -> Pos.none (P_Appl (type_to_pattern t1, type_to_pattern t2))
+  | Prod _       -> fatal None "Prod - Not yet implemented"
+  | Abst _       -> fatal None "Abst - Not yet implemented"
+  | LLet _       -> fatal None "LLet - Not yet implemented"
+  | Type         -> fatal None "TYPE - Not yet implemented"(*Pos.none P_Type*)
+  | Kind         -> fatal None "[Kind] not possible"
+  | Wild         -> fatal None "[Wild] not possible"
+  | TRef _       -> fatal None "[TRef] not possible"
+  | Meta _       -> fatal None "[Meta] not possible"
+  | Patt _       -> fatal None "[Patt] not possible"
+  | TEnv _       -> fatal None "[TEnv] not possible"
+
+(** Some constructors to create a p_term without position *)
+
+let _P_Iden : qident -> bool -> p_term = fun ident b ->
```

Comment:
thanks to rename into p_iden, p_patt, etc.

Reply:
We could also use a module P with functions iden, patt, app, fapp, and then use P.iden, P.patt, etc.

Reply:
`let iden : string -> p_term = fun s -> Pos.none (P_Iden(Pos.none([],s),false))`

Reply:
Why the boolean is false?

Reply:
right, you can set it to true

Reply:
In fact, I don't the semantic of this boolean. I believe it's not explain everywhere in the doc.

Reply:
syntax.ml is not very clear about this indeed:
```
  | P_Iden of qident * bool
  (** Variable or (qualified, explicitly applied) symbol. *)
```
You get true when the identifier is prefixed by `@` meaning that implicit arguments must be explicitly given. See parser.ml:
```
let parser term_ident =
  | expl:{"@" -> true}?[false] qid:qident -> in_pos _loc (P_Iden(qid, expl))
```

Reply:
No no everything it's OK. You can read this in `syntax.ml`:
```
(** {b NOTE} the boolean parameter of {!constructor:P_Iden} tells whether  the
    corresponding symbol is explicitly applied (value [true]) or not. The flag
    hence indicates whether the symbol has been prefixed with ["@"]. *)
```

Reply:
fixed in 3537d41e0fe9fdab85eb9e42e57d4dff55e37da5